### DISCUSSION TITLE
[Bug]: HTMLInputElement.files setter: Value being assigned does not implement interface FileList.

### DISCUSSION THREAD

**Discussion by jsefiani** [react-hook-form/react-hook-form#6649] [Q&A]
### Version Number

7.15.3

### Codesandbox/Expo snack

https://codesandbox.io/s/condescending-feather-57ded?file=/src/App.js

### Steps to reproduce

Click on "Choose file" to upload a photo (it directly uploads your file and saves the response of the API call to the form state)

### Expected behaviour

When uploading the file, it should save the response of the API call (id and URL of the resource) to the form state such that I can save it together with the whole form and create a product.

Desired form state:
```
{
 productName: "",
 file: {
  id: "",
  url: ""
 }
}
```

### What browsers are you seeing the problem on?

Firefox, Chrome, Safari, Edge

### Relevant log output

```shell
When I try to save the the API response from the uploaded file to the form state, I get the following type error:

HTMLInputElement.files setter: Value being assigned does not implement interface FileList.
```

I don't want to save the file to the form state, as that is not needed in my setup but nevertheless `setValue` is expecting the value to implement the `FileList` interface while I just want to save the ID an URL of the uploaded file.

Let me know if anything is unclear.
```


### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by bluebill1049** [comment]
FileList object is immutable. What are you try to achieve here?

**Reply by jsefiani** [reply]
I'm trying to independently upload a file and save its response (ID and URL of uploaded file) to the form state, such that it can be saved together with the product.

But it doesn't allow me to do so because of that type error. Why is RHK assuming that file's value should be of type `FileList`?

This should ideally be saved when clicking on `Save Product`:
```
{
  productName: "name",
  file: {
    id: "sdsds",
    url: "link to AWS S3 object"
  }
}
```

**Reply by jsefiani** [reply]
This is my `onChange` handler for the file input:
```
  const handleLogoUpload = (e: React.ChangeEvent<HTMLInputElement>) => {
    const file = e.target.files?.[0] as File;

    uploadFileMutation(file, {
      onSuccess: (response) => {
        // results in a type error as RHK is expecting response to be an instance of FileList while it isn't
        formUtils.setValue('logo', response); // response => {id: string, url: string}
      },
    });
  };
```

**Reply by bluebill1049** [reply]
you can not use setValue for FileList object, if you need to show image, save into a local state and update it as string.

**Reply by jsefiani** [reply]
I don't want to save a FileList object nor do I want to show an image, all I want to do is to save the following object: `{id: string, url: string}` to the form state using `setValue`.

But RHF doesn't allow me to do so? Don't you think this is a bug or is this expected?

Please have a look at my sandbox link, it's pretty clear what I'm trying to achieve.

**Reply by jsefiani** [reply]
Let me know if you need any other details.

**Reply by bluebill1049** [reply]
what I meant is not to use `setValue` for that, use the following instead.

```
const [path, setPath]= React.useState('path')

setPath('with the url');

<img src={path} />
```

if you want to upload image `<input type="file" {...register('file')} />` that will do, and for display image with URL use local state.
